### PR TITLE
Adding remote config/logging capabilities to Windows build

### LINKED ISSUE #2425: Windows: run python-based integration tests (including TLS server)
There are several tests that use Python to start HTTPS servers and assure the TLS-based plugins are working correctly. These need to be working for Windows. Most of the work here is assuring CMake can launch Python and run these tests (report the results back).

### LINKED ISSUE #2426: Windows: enable remote and TLS plugins in CMake
Remove this: https://github.com/facebook/osquery/blob/master/osquery/CMakeLists.txt#L149

### PR DESCRIPTION
**Pull request by yying**
Enabled remote config and logging capabilities using TLS. 

For remote TLS logging and/or filesystem logging to work correctly, a glog with a bug fix (https://github.com/google/glog/pull/123).

Closes #2425 
Closes #2426

### CODE REVIEW

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by osqueryer** [comment]
Can one of the osquery admins reply with "ok to test" to kick off a build...

**Comment by facebook-github-bot-3** [comment]
@yying updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2469/files/c7923535e6649153628f4b68e782085d0a6514f0..c0b1a61c0ce4d70225874b0cb6272c0ff9f530b0)

**Comment by facebook-github-bot-5** [comment]
@yying updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2469/files/c0b1a61c0ce4d70225874b0cb6272c0ff9f530b0..8ba6b7a0dc31411078e1414bdb24f6c27a0af381)

**Comment by mtmcgrew** [comment]
ok to test

**Comment by osqueryer** [comment]
:no_good: The commit aa121edad2757774bd02cc491ca112db8527bdbc (Job results: [238](https://jenkins.osquery.io/job/osqueryPullRequestAudit/238/)) failed the code audit.

**Comment by osqueryer** [comment]
:-1: The commit aa121edad2757774bd02cc491ca112db8527bdbc (Job results: [265](https://jenkins.osquery.io/job/osqueryPullRequestBuild/265/)) failed one or more tests (**Windows**).

**Comment by facebook-github-bot-7** [comment]
@yying updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2469/files/aa121edad2757774bd02cc491ca112db8527bdbc..bb99bff57306eba60d3b31ec584f53c63e779a1c)

**Comment by osqueryer** [comment]
:-1: The commit bb99bff57306eba60d3b31ec584f53c63e779a1c (Job results: [268](https://jenkins.osquery.io/job/osqueryPullRequestBuild/268/)) failed one or more tests (**Windows**).

File: osquery/core/windows/utils.cpp [hunk 1]
```diff
@@ -27,7 +29,9 @@ std::string platformAsctime(const struct tm* timeptr) {
     return "";
   }
 
-  return std::string(buffer.data(), buffer.size());
+  std::string time(buffer.data());
```
Comment (theopolis) [inline osquery/core/windows/utils.cpp]: nit, avoid `time` as a local variable name.

File: osquery/remote/tests/requests_tests.cpp [hunk 2]
```diff
@@ -129,13 +133,32 @@ TEST_F(RequestsTests, test_compression) {
...
+  /*
+   * gzip header has a field that specifies the filesystem the compression took
+   * place, we need to separate this for NTFS and Unix
+   *
+   * Reference: http://www.zlib.org/rfc-gzip.html
+   */
+  std::string expected;
+  if (isPlatform(PlatformType::TYPE_WINDOWS)) {
+    expected = std::string(
+        "\x1F\x8B\b\0\0\0\0\0\x2\v\xED\xC4\xB1\r\0\0\x4\0\xB0s\xC5"
```
Comment (theopolis) [inline osquery/remote/tests/requests_tests.cpp]: nit there's only 1 character difference? Maybe we can use some concatenation?
```cpp
std::string expected = "\x1F\x8B\b\0\0\0\0\0\x2";
expected += (isPlatform(PlatformType::TYPE_WINDOWS)) ? "\v" : "\x3";
expected += ...
```

**Comment by facebook-github-bot-7** [comment]
@yying updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2469/files/bb99bff57306eba60d3b31ec584f53c63e779a1c..8358e67eb9d2019d4a34a466a133b412f35e1803)

**Comment by osqueryer** [comment]
:-1: The commit 8358e67eb9d2019d4a34a466a133b412f35e1803 (Job results: [3263](https://jenkins.osquery.io/job/osqueryPullRequestBuild/3263/)) failed one or more tests (**OS X/Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 8358e67eb9d2019d4a34a466a133b412f35e1803 (Job results: [273](https://jenkins.osquery.io/job/osqueryPullRequestBuild/273/)) failed one or more tests (**Windows**).

**Comment by facebook-github-bot-3** [comment]
@yying updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2469/files/8358e67eb9d2019d4a34a466a133b412f35e1803..a150ba9e4a43f3887a8c36b55f91589b8772b39a)

**Comment by osqueryer** [comment]
:-1: The commit a150ba9e4a43f3887a8c36b55f91589b8772b39a (Job results: [280](https://jenkins.osquery.io/job/osqueryPullRequestBuild/280/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit a150ba9e4a43f3887a8c36b55f91589b8772b39a (Job results: [3270](https://jenkins.osquery.io/job/osqueryPullRequestBuild/3270/)) failed one or more tests (**OS X/Linux**).

**Comment by theopolis** [comment]
If you rebase this PR the CentOS7 error will disappear.

**Comment by theopolis** [comment]
Error on Windows:
```
C:\ProgramData\chocolatey\lib\cpp-netlib\local\include\boost/network/uri/builder.hpp(9):\
 fatal error C1083: Cannot open include file: 'asio/ip/address.hpp': No such file or directory 
```

Is ASIO not installed? `cpp-netlib` uses the non-boost ASIO implementation, I'm not sure what modifications/version-lag boost has at this point, but this may be the source of the error.

**Comment by yying** [comment]
I believe what may have happened is that the `cpp-netlib` being used at the moment on Jenkins is outdated. During this PR, I changed around the include paths within in `cpp-netlib` chocolatey package. I think forcing a `cpp-netlib` should do the trick...

**Comment by facebook-github-bot-7** [comment]
@yying updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2469/files/a150ba9e4a43f3887a8c36b55f91589b8772b39a..9708281f5a57c8dc0a2afff10167b954d29e738e)

**Comment by facebook-github-bot-8** [comment]
@yying updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2469/files/9708281f5a57c8dc0a2afff10167b954d29e738e..b53eeeb5df1309620863f6c3f8c1b13cb28e4511)

**Comment by osqueryer** [comment]
:-1: The commit b53eeeb5df1309620863f6c3f8c1b13cb28e4511 (Job results: [291](https://jenkins.osquery.io/job/osqueryPullRequestBuild/291/)) failed one or more tests (**Windows**).

**Comment by facebook-github-bot-9** [comment]
@yying updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2469/files/b53eeeb5df1309620863f6c3f8c1b13cb28e4511..2cea2f205718d4d711fdc294cd630c4a62fec859)

File: osquery/remote/transports/tls.cpp [hunk 3]
```diff
@@ -105,8 +111,15 @@ http::client TLSTransport::getClient() {
...
     } else {
       // There is a non-default server certificate set.
+      boost::system::error_code ec;
+
+      auto status = fs::status(server_certificate_file_, ec);
       options.openssl_verify_path(server_certificate_file_);
-      options.openssl_certificate(server_certificate_file_);
+
+      // On Windows, we cannot set openssl_certificate to a directory
+      if (status.type() == fs::regular_file) {
```
Comment (theopolis) [inline osquery/remote/transports/tls.cpp]: We should include an `else` and a `LOG(WARNING)`.

File: osquery/remote/transports/tls.h [hunk 4]
```diff
@@ -10,6 +10,40 @@
 
 #pragma once
 
+// Our third-party version of cpp-netlib uses OpenSSL APIs.
```
Comment (theopolis) [inline osquery/remote/transports/tls.h]: Maybe `/*`.

File: osquery/remote/transports/tls.h [hunk 5]
```diff
@@ -10,6 +10,40 @@
...
+// This is placed here because of ordering issues. ASIO requires WinSock.h
+// not to be already included.
+// clang-format off
+#ifdef WIN32
+#pragma warning(push, 3)
+
+/// Suppressing warning C4005: 'ASIO_ERROR_CATEGORY_NOEXCEPT': macro redefinition
+#pragma warning(disable: 4005)
+
+// Suppressing warning C4244: 'argument': conversion from '__int64' to 'long', possible loss of data
```
Comment (theopolis) [inline osquery/remote/transports/tls.h]: nit, `///` for consistency.

File: osquery/remote/transports/tls.h [hunk 6]
```diff
@@ -10,6 +10,40 @@
...
+#pragma clang diagnostic ignored "-Wunused-local-typedef"
+#pragma clang diagnostic ignored "-W#pragma-messages"
+#endif
+
+#include <boost/network/protocol/http/client.hpp>
+
+#ifdef WIN32
+#pragma warning(pop)
+
+// We need to reinclude this to re-enable boost's warning suppression
```
Comment (theopolis) [inline osquery/remote/transports/tls.h]: nit, same

**Comment by theopolis** [comment]
Ok, one last round of changes and this looks good.

**Comment by facebook-github-bot-9** [comment]
@yying updated the pull request - [view changes](https://github.com/facebook/osquery/pull/2469/files/2cea2f205718d4d711fdc294cd630c4a62fec859..235222008d5c3766dad8a24523e2cc614d17e0fe)

**Pull request merged**